Add webhook which reject storage class name change

martindekov opened the pull request:
Adding webhook which rejects storage class name changes as part of the `harvesterhci.io/volumeClaimTemplates` annotation.

Creating, deleting and adding to the annotaion is allowed, but changing the storage class name itself is rejected. Also setting empty storage class name is allowed as well - following the omitempty flag of the field in the object.

#### Problem:
Changes to the volume's storage class is rejected through UI, but allowed through API - direct edits of the raw yaml.

#### Solution:
Introduce webhook which will detect such changes to already existing attached volumes and reject changes to the storage class name.

#### Related Issue(s):
Issue #8772

#### Test plan:
1. Build iso with the change
2. Create local harvester with the iso (single node)
3. Create storage class - `default-sc` with replication 1 and set it as default storage class
4. Create VM
5. Create volume and attach it as additional volume to the VM
6. Try to edit the original attached volume through kubectl and UI the value of key `storageClassName` - expect rejection (need to scroll a little):
```
    harvesterhci.io/volumeClaimTemplates: >-
      [{"metadata":{"name":"test-disk-0-kgsuo","creationTimestamp":null,"annotations":{"harvesterhci.io/imageId":"default/image-wd7kd"}},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"10Gi"}},"storageClassName":"default-sc","volumeMode":"Block"},"status":{}},{"metadata":{"name":"testtest","creationTimestamp":null},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"10Gi"}},"storageClassName":"default-sc","volumeMode":"Block"},"status":{}}]
                                                                                                                                                                                                                                                                                                                                                                                                                                              ### from this here ###########
```
             
to (need to scroll a little):
```
    harvesterhci.io/volumeClaimTemplates: >-
      [{"metadata":{"name":"test-disk-0-kgsuo","creationTimestamp":null,"annotations":{"harvesterhci.io/imageId":"default/image-wd7kd"}},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"10Gi"}},"storageClassName":"default-sc","volumeMode":"Block"},"status":{}},{"metadata":{"name":"testtest","creationTimestamp":null},"spec":{"accessModes":["ReadWriteMany"],"resources":{"requests":{"storage":"10Gi"}},"storageClassName":"longhorn","volumeMode":"Block"},"status":{}}]
                                                                                                                                                                                                                                                                                                                                                                                                                                              ### to this here ###########
```
7. Try to edit the additional attached volume as above

#### Additional documentation or context
Initially logic was not triggered because of [this check](https://github.com/harvester/harvester/blob/abce991f78d6c4020cf2dff6f72c660d7124e098/pkg/webhook/resources/virtualmachine/validator.go#L268) always returning nil:
```
	if oldVM.Spec.Template != nil && newVM.Spec.Template != nil && reflect.DeepEqual(oldVM.Spec.Template.Spec.Domain.Devices.Interfaces, newVM.Spec.Template.Spec.Domain.Devices.Interfaces) {
		return nil
	}
```
Added a comment above it so all new validation should come before the comment and the check.

martindekov committed:
Add webhook which reject storage class name change

Adding webhook which rejects storage class name changes as part
of the harvesterhci.io/volumeClaimTemplates annotation.

Creating, deleting and adding to the annotaion is allowed, but
changing the storage class name itself is rejected.

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

w13915984028 reviewed (dismissed):
LGTM, thanks for the fix.

martindekov committed:
Address feedback from Ivan

Addressing feedback from Ivan to merge the new function with
already existing one - checkResizeVolumes as the major part
of the logic of parsing json etc was already covered.

Renamed the method from `checkResizeVolumes` to
`checkVolumeAnntations` to better reflect what the
method is doing

Signed-off-by: Martin Dekov <martin.dekov@suse.com>

martindekov committed:
Address feedback from Olivier and Jian

Addressing feedback from Olivier and Jian including the following:
* Check for nil poiners before comparison to avoid panics fixing
the integration test
* Augment error message so we include all attempted name changes
of the storage class
* Add test to validate the empty value works as opposed to panicking
and that we allow such values

Signed-off-by: Martin Dekov <martin.dekov@suse.com>
